core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java

```diff
@@ -0,0 +1,170 @@
+    private final String mTemporaryPath;
+
+    private UnderFileSystemOutputStream(String ufsPath, Configuration conf)
+        throws FileAlreadyExistsException, IOException {
+      mConf = Preconditions.checkNotNull(conf);
+      mPath = Preconditions.checkNotNull(ufsPath);
+      mTemporaryPath = PathUtils.temporaryFileName(IdUtils.getRandomNonNegativeLong(), mPath);
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      if (ufs.exists(mPath)) {
+        throw new FileAlreadyExistsException(ExceptionMessage.FAILED_UFS_CREATE.getMessage(mPath));
+      }
+      mStream = ufs.create(mTemporaryPath);
+    }
+
+    /**
+     * Closes the temporary file and deletes it. This object should be discarded after this call.
+     *
+     * @throws IOException if an error occurs during the under file system operation
+     */
+    private void cancel() throws IOException {
+      mStream.close();
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      // TODO(calvin): Log a warning if the delete fails
+      ufs.delete(mTemporaryPath, false);
+    }
+
+    /**
+     * Closes the temporary file and attempts to promote it to the final file path. If the final
+     * path already exists, the stream is canceled instead.
+     *
+     * @throws IOException if an error occurs during the under file system operation
+     */
+    private void complete() throws IOException {
+      mStream.close();
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      if (!ufs.rename(mTemporaryPath, mPath)) {
+        // TODO(calvin): Log a warning if the delete fails
+        ufs.delete(mTemporaryPath, false);
+      }
+    }
+
+    /**
+     * @return the wrapped output stream to the under file system file
+     */
+    private OutputStream getStream() {
+      return mStream;
+    }
+  }
+
+  /** A random id generator for worker file ids. */
+  private AtomicLong mIdGenerator;
+  /** Map of worker file ids to open under file system streams. */
+  private ConcurrentMap<Long, UnderFileSystemOutputStream> mStreams;
+
+  /**
+   * Creates a new under file system manager. Stream ids are unique to each under file system
+   * manager.
+   */
+  public UnderFileSystemManager() {
+    mIdGenerator = new AtomicLong(IdUtils.getRandomNonNegativeLong());
+    mStreams = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Creates a {@link UnderFileSystemOutputStream} for the given file path and adds it to the
+   * open streams map keyed with the returned worker file id.
+   *
+   * @param ufsPath the path to create in the under file system
+   * @return the worker file id which should be used to reference the open stream
+   * @throws FileAlreadyExistsException if the under file system path already exists
+   * @throws IOException if an error occurs when operating on the under file system
+   */
+  public long createFile(String ufsPath) throws FileAlreadyExistsException, IOException {
```

Comment:
are these paths just the path component, or the full URI in String form? The way we use uri and path interchangeably is sometimes confusing